Simulcast

Sean-Der opened the pull request:
Add WebRTC Simulcast Support

mpisat commented:
How can we test this? I downloaded latest build from artifacts but I don't see any new option on OBS? A little help? :)

Sean-Der commented:
Thanks for trying it out @mpisat!

I described how to use it [here](https://github.com/obsproject/rfcs/pull/55#issuecomment-1627791991) 

I am adding support for Advanced Output Simulcast today. I will make sure to post updates on this PR as I work on it.

alexthedamager commented:
Tested with Broadcast Box - works great! This is very cool indeed.

ygit commented:
Nice! 🥇

chhofi commented:
Hey @Sean-Der thanks for your great work! I am running the Arm MacOs OBS version and can successfully stream with simulcast enabled to the website. At least that is what I think is happening as all the metrics look good in OBS. On the website I also get the simulcast controls but unfortunately the stream does not load whether I use Safari or Chromium as a browser. What is the best way to debug this further and provide feedback?

SetoKaiba commented:
Hi @Sean-Der , I got the same problem as @chhofi .
I tried it with siobud. And this is my settings.
![image](https://github.com/Sean-Der/obs-studio/assets/159540/be5ff73d-5093-413e-8f94-f00d921a2197)
It's not loading.
![image](https://github.com/Sean-Der/obs-studio/assets/159540/e34619ef-d221-40ac-9f37-1977b9aace7b)
It does show the quality levels.
![image](https://github.com/Sean-Der/obs-studio/assets/159540/415f61c2-50e1-4366-b5a7-5890a3d73dd0)
When I click stop, it shows this. Is it intended?
![image](https://github.com/Sean-Der/obs-studio/assets/159540/0611896e-9e02-4783-b6e8-1836fedee20f)

Sean-Der commented:
Sorry about that @SetoKaiba @chhofi 

What encoder are using? Can you try these settings? Does it help if you lower the bitrate? Does it work without Simulcast enabled?


![Screenshot 2023-08-05 at 2 48 31 PM](https://github.com/Sean-Der/obs-studio/assets/1302304/6799fc90-28a4-4561-97bb-2bcd351f6312)

SetoKaiba commented:
Hi, @Sean-Der .
I use NVIDIA NVENC H.264 before.
I tried with your settings. It's sometimes able to load with green or blur pixels and extra low framerate, maybe 1-5.
But I can't determine whether this is caused by network.

And then I tried it with my local ZLMediaKit server. It's working with OBS WHIP before. It supports simulcast.
I tried with the translate-libdatachannel-to-cpp branch. It's working without simulcast.
I tried with the simulcast branch. It's not working even without simulcast.

And here's the warning I saw from ZLMediaKit logs for simulcast branch.
I didn't enable simulcast.
```
2023-08-06 16:34:53.251 W [MediaServer] [667-event poller 0] WebRtcTransport.cpp:779 onRtcp | 未识别的sr rtcp包:version:2
padding:0
report_count:0
pt:200(RTCP_SR)
size:28
--------
ssrc:1043019512
ntpmsw:3900299693
ntplsw:1079742464
ntp time:2023-08-06 16:34:53.251
rtpts:3000
packet_count:0
octet_count:0
```
"未识别的sr rtcp包" means "unrecognized sr rtcp packet". 

~~And also, I found that the stream is registered as _l _m _h in ZLMediaKit even I didn't enable simulcast. Is the checkbox working?~~
It seems the change of simulcast checkbox need restart of OBS. May be it's a bug.

SetoKaiba commented:
![image](https://github.com/Sean-Der/obs-studio/assets/159540/26ebdaa7-5049-4354-96df-61362760806f)
Here's the screenshot for using your settings with your server. 
And also, I checked the chrome://webrtc-internals with choosing the quality level.
It someimes can be switched to other level. I have never switched to h level.
![image](https://github.com/Sean-Der/obs-studio/assets/159540/a3678fcd-3f37-4c44-aad9-5807323f0a8b)

Sean-Der commented:
@SetoKaiba can you try running https://github.com/glimesh/broadcast-box locally?

Just to debug
* is issue OBS encoding 
* Is issue network 
* Is issue WebRTC implementation

SetoKaiba commented:
@Sean-Der I tried with broadcast-box `go run .` locally. It's working without problem.

Another small question, I run it with docker in windows. The video is not loaded. But the streaming doesn't report any error.

For ZLMediaKit, I'll try to debug the problem with ZLMediaKit's maintainers. But the ZLMediaKit is working with simulcast. Its demo page run successfully. So it seems to be a compatiblity problem.

Thank you.

Sean-Der commented:
@SetoKaiba thanks for all the testing!

My guess is that we aren't implementing something properly? If you can figure out what ZLMediaKit is looking for I can implement that right away!

SetoKaiba commented:
@Sean-Der The ZLMediaKit author is on vocation. So I have a look at it.

![image](https://github.com/Sean-Der/obs-studio/assets/159540/f060da0f-9f8a-4f3a-aee5-280788ae7908)
It seems that the extesion length is wrong.

And also, it doesn't consider the length of mid or rid is longer than one char. If it's more that one char, it should be 4 byte aligned. But it's ok now since that it's that mid is "1" and rid is "h", "m", "l".
![image](https://github.com/Sean-Der/obs-studio/assets/159540/1a205dc0-fbe9-418a-9d5b-389942cae770)

SetoKaiba commented:
```C++
        size_t rtpExtHeaderSize = 4; // 0xbede & 2byte length
        size_t headerLength = 0;

        if (setVideoRotation) {
		rtpExtHeaderSize += 2;
		headerLength +=1;
	}

	if (rtpConfig->mid.has_value()) {
		rtpExtHeaderSize += (1 + rtpConfig->mid->length());
		headerLength +=1;
	}

	if (rtpConfig->rid.has_value()) {
		rtpExtHeaderSize += (1 + rtpConfig->rid->length());
		headerLength +=1;
	}
        size_t padding = (4 - (rtpExtHeaderSize % 4)) % 4;
	rtpExtHeaderSize += padding;
```
```C++
		extHeader->setHeaderLength(headerLength);
```
Maybe it should be like this.

SetoKaiba commented:
@Sean-Der I test it with SRS, ZLMediaKit. They don't work because of the extension header serialization problem.

SetoKaiba commented:
@Sean-Der Finally find out the problem with ZLMediaKit. The ZLMediaKit doesn't support mid. But this Simulcast implementation will always send the mid. He claims that mid is not part of the spec of Simulcast. The streaming should follow the answer.

Sean-Der commented:
@Sean-Der if you delete this line https://github.com/Sean-Der/obs-studio/pull/2/files#diff-ae779f158b6d305fd42fd223123fa37e08cbd282a8fd479fc17447eff7c65a16R205 (and don't send MID) Simulcast then works for you?

SetoKaiba commented:
> @Sean-Der if you delete this line https://github.com/Sean-Der/obs-studio/pull/2/files#diff-ae779f158b6d305fd42fd223123fa37e08cbd282a8fd479fc17447eff7c65a16R205 (and don't send MID) Simulcast then works for you?

Yes. Or not using the directProxy of the ZLMediaKit, it will reassemble the rtp packet without extension header. Then it will work. He did a small fix now to filter the ext by answer instead of offer. It works also.
Anyway, he said that the packetization should follow the sdp exchange answer.

Sean-Der committed:
obs-webrtc: Add Simulcast Support

mpisat commented:
https://obsproject.com/kb/whip-streaming-guide shows screenshots with simulcast, yet this was never added to OBS.
When we'll have Sean's simulcast added to OBS? I mean, I don't mind downloading an older build of OBS to use that feature but authors of OBS either add that feature themselves or allow PR...

Sean-Der commented:
Sorry for the delay @mpisat!

PR is open against OBS upstream here https://github.com/obsproject/obs-studio/pull/10885 if you want to test.

Other higher priority issues are getting addressed first. I am not sure where on the OBS roadmap Simulcast falls. Would it be useful if I rebased this PR (so it is on top of latest OBS?)

mpisat commented:
A rebase would be great if you have time. I can also test latest build once it's done.
I commented on https://github.com/obsproject/obs-studio/pull/10885 and fingers crossed.

Warchamp7 committed:
frontend: Adjust application shutdown logic

Improves app shutdown in a few ways, including separating out different
pieces of the OBSBasic close handler into their own functions.

Removes the crash handler sentinel earlier when the main window is closed,
preventing unclean shutdown warnings when a plugin causes issues. While not
ideal, the dialog is not useful when we cannot specify which plugin caused the
problem.

Increases shutdown priority of the main application so that when OBS interrupts
the session ending, CEF is not closed at the same time. This fixes a crash.

Additional safeguards and event handling to try to ensure a smoother shutdown.

dsaedtler committed:
frontend: Fix recursion during canvas removal causing crashes

tytan652 committed:
CI: Prepare more space for Flatpak jobs

pkviet committed:
libobs: Add devices_match to null monitor

This adds the devices_match function to null monitor to fix linking
issues on linux when pulse audio is disabled.
Fixes #12810

Signed-off-by: pkv <pkv@obsproject.com>

Warchamp7 committed:
cmake: Fix order of frontend targets

Updates targets to be consistent in their order putting uppercase before lowercase

Warchamp7 committed:
frontend: Remove crash handler earlier in shutdown

Warchamp7 committed:
docs: Add API disclaimer to canvases

RytoEX committed:
libobs: Update version to 32.0.3

exeldro committed:
libobs: Do not wait for audio of removed sources

pkviet committed:
libobs: Improve monitoring deduplication

The monitoring deduplication was previously checking at each audio tick
the whole audio tree for Audio Output Capture (AOC) devices used for
monitoring. The profiling did not show any big impact on the audio
callback. But due to reports of a significant slowdown for scenes with
numerous audio sources, we have moved the check on AOC from the audio
thread to the UI thread.

So we implemented obs_source_audio_output_capture_device_changed which
is triggered whenever:
- a monitoring device is changed in Settings > Audio,
- an Audio Output Capture source changes its device or on startup.

This function compares the AOC device with the monitoring device in UI
thread. If they are identical, a signal is sent to the audio thread to
add an audio task updating the AOC duplicating source pointer at the end
of an audio tick.

This triggers monitoring deduplication if the ptr is not NULL.
The calls in the AOC are implemented in next commits.
The rest of the logic in obs-audio.c is the same except on one count,
which is that we check against the muted state of the AOC rather than
its user_muted; with the new logic, muted works better.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
win-wasapi: Add monitoring deduplication calls

For 'wasapi output capture', calls are added when devices are either
changed, started or removed.

The calls trigger checks against the monitoring device.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
mac-capture: Add monitoring deduplication calls

For 'coreaudio_output_capture', calls are added when devices are either
changed, started or removed.

The calls trigger a check against the monitoring device.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
linux-pulseaudio: Add monitoring deduplication calls

For 'pulse_output_capture', calls are added when devices are either
changed, started or removed.

The calls trigger a check against the monitoring device.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
libobs: Fix default macOS monitoring device

The default device retrieved was the one used for system sounds.
This fixes it by retrieving the one used to output audio.

It seems the bug has been there since the initial writing of this bit of
code (so 2017).

Warchamp7 committed:
frontend: Switch off deleted scenes immediately

Warchamp7 committed:
frontend: Fix transition duration not loading correctly

RytoEX committed:
libobs: Update version to 32.0.4
